### PR TITLE
Support lang pseudo class

### LINKED ISSUE #8219: Implement `:lang` CSS pseudo-class
Spec: http://www.w3.org/TR/selectors/#lang-pseudo .  Information about how to actually determine the language of an element in HTML: https://html.spec.whatwg.org/multipage/dom.html#language .  The XML specification has similar language: http://www.w3.org/TR/2008/PER-xml-20080205/#sec-lang-tag .

Requires some changes to https://github.com/servo/rust-selectors so it can parse `:lang`.

Should be straightforward to implement.

### PR DESCRIPTION
**Pull request by KiChjang**
Fixes #8219.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/15464)

### CODE REVIEW

**Review by emilio — CHANGES_REQUESTED** [review]
r-, you can't call a method borrowing a refcell from style code, since it may race. We'll need to figure another way of doing this.

(Also, probably `lang` should be an `Atom`).

File: components/script/dom/element.rs
Comment (emilio) [inline components/script/dom/element.rs]: You need to call get_attr_for_layout instead (if you're in layout).

**Comment by jdm** [comment]
You'll want to enable the tests in css-tests/selectors-3_dev, since there are a bunch of tests for :lang there.

File: components/script/dom/element.rs
Comment (emilio) [inline components/script/dom/element.rs]: Also, I have a pretty important question here... Where is this invalidated?

Comment (KiChjang) [inline components/script/dom/element.rs]: As in when is this being set to None? When it is first initialized. Doesn't make much sense to do a tree traversal on creation of the element, yes?

Comment (jdm) [inline components/script/dom/element.rs]: The question is what happens if the lang attribute is updated after it's cached?

Comment (emilio) [inline components/script/dom/element.rs]: Right

Comment (emilio) [inline components/script/dom/element.rs]: To be precise, not only if the lang attribute is cached, but also one of the other conditions that could make lang change its value.

**Comment by SimonSapin** [comment]
As noted in https://github.com/servo/servo/issues/8219#issuecomment-277878293, Gecko doesn’t do any caching of the language on elements. Instead, it will traverse up the tree as necessary to resolve `:lang()` during selector matching.

So, I don’t know if my untested idea to cache is good after all?

File: components/script/dom/element.rs
Comment (KiChjang) [inline components/script/dom/element.rs]: Ah right, that is supposed to be invalidated if the attribute's value is mutated, and should be handled in `attribute_mutated`. My current PR still doesn't have any of those changes, but will push a fix soon.

Comment (emilio) [inline components/script/dom/element.rs]: I believe it's not worth to cache given the complexity of the cases, it's not only when the attribute was mutated, and given it is such an uncommon attribute _and_ selector.

File: components/script/dom/element.rs
Comment (KiChjang) [inline components/script/dom/element.rs]: The WPT tests are still giving me TIMEOUT errors, and I suspect it's something to do with being deadlocked. Can somebody confirm that the temporary `RwReadGuard` is dropped immediately after this conditional?

Comment (emilio) [inline components/script/dom/element.rs]: Not sure, but as I said, you can't just call `get_attribute` in a method that get's called for layout, that's plain racy.

Comment (KiChjang) [inline components/script/dom/element.rs]: Yes, but the layout code doesn't call this function anymore, `fn language()` for `LayoutJS<Element>` now does its own attribute/parent node traversal.

File: components/script/dom/element.rs
Comment (KiChjang) [inline components/script/dom/element.rs]: It also bloats the size of the element struct, so perhaps not caching would be the best way to go here.

File: components/script/dom/element.rs
Comment (emilio) [inline components/script/dom/element.rs]: I'm somewhat concerned about this duplication, can we factor this out somehow?

File: components/script/dom/element.rs
Comment (emilio) [inline components/script/dom/element.rs]: Probably call this `get_lang_for_layout` or similar so it's more obvious that you're calling the correct method below.

File: components/script/dom/element.rs
Comment (emilio) [inline components/script/dom/element.rs]: I'd avoid to do this recursively TBH, could we use a iterator/loop instead? It seems that otherwise creating a very deep dom could overflow the stack fairly easily.

File: components/style/servo/selector_parser.rs
Comment (emilio) [inline components/style/servo/selector_parser.rs]: This seems somewhat flacky. also, I believe you need to escape the language (see `CSSStringWriter`. Verify that other browsers escape it and please add a test for this to WPT if there isn't any already.

Instead of this dummy variable, can't you do something like returning early in the `Lang` case?

File: components/style/servo/selector_parser.rs
Comment (emilio) [inline components/style/servo/selector_parser.rs]: nit: Align this like so:

```rust
fn parse_non_ts_functional_pseudo_class(&self,
                                        name: Cow<str>,
                                        parser: &mut CssParser)
                                        -> Result<NonTSPseudoClass, ()> {
```

**Review by emilio — CHANGES_REQUESTED** [review]

File: components/script/dom/element.rs
Comment (KiChjang) [inline components/script/dom/element.rs]: This is super duper unpossible. I used many methods to try and make this generic but to no avail. The core problem here is the `LayoutJS<T>` pointers that needs to constantly be interchanged between `JS<T>`.

File: components/style/servo/selector_parser.rs
Comment (KiChjang) [inline components/style/servo/selector_parser.rs]: `CSSStringWriter` doesn't exist, which method/struct are you talking about?

Comment (emilio) [inline components/style/servo/selector_parser.rs]: `CssStringWriter`, from cssparser, sorry for the confusion

**Comment by KiChjang** [comment]
Comments addressed.

@bors-servo try

**Comment by jdm** [comment]
Don't forget enabling css-tests/selectors-3_dev!

**Comment by KiChjang** [comment]
Okay, I think I got all the test expectations correct this time. r? @emilio

File: components/style/servo/selector_parser.rs
Comment (emilio) [inline components/style/servo/selector_parser.rs]: Any reason to not use just `unreachable!` here?

File: components/style/servo/selector_parser.rs
Comment (emilio) [inline components/style/servo/selector_parser.rs]: I still believe this should be an atom, but if you don't agree, let's at least use `Lang(Box<str>)`, so the size increment is not so huge.

File: components/script/dom/element.rs
Comment (emilio) [inline components/script/dom/element.rs]: when can a non-element be a parent of an element? Let's just make `current_node` be `current_element`, and use something like `node.parent_node_ref().and_then(|p| p.downcast::<Element>())`, or something like that, breaking when it's none.

File: components/script/dom/element.rs
Comment (emilio) [inline components/script/dom/element.rs]: nit: (optional), I think using the pattern:

```rust
if foo {
    return;
}
if bar {
    return
}
```

Is more easy to read forward than:

```rust
if foo {
    return;
} else if bar {
    return;
}
```

But your call, I won't insist on this.

File: components/script/dom/element.rs
Comment (emilio) [inline components/script/dom/element.rs]: nit: I believe you can use `String::from(&attr)`.

File: tests/wpt/metadata-css/selectors-3_dev/html/css3-selectors-lang-009.htm.ini
Comment (emilio) [inline tests/wpt/metadata-css/selectors-3_dev/html/css3-selectors-lang-009.htm.ini]: I guess i should ask about all these failing tests? i know there were skipped before, but is there a problem in implementing this stuff right? if so, mind filing issues?

File: tests/wpt/metadata/dom/nodes/ParentNode-querySelector-All.html.ini
Comment (emilio) [inline tests/wpt/metadata/dom/nodes/ParentNode-querySelector-All.html.ini]: Similarly. I guess this is about the extended filtering `FIXME`? If so, please file an issue.

File: components/style/servo/selector_parser.rs
Comment (KiChjang) [inline components/style/servo/selector_parser.rs]: Sooner or later we're going to have to implement https://tools.ietf.org/html/rfc4647#section-3.3.2, storing a String here would just ease the implementation of it. Not sure if a `Box<str>` would achieve the same effect.

File: components/style/servo/selector_parser.rs
Comment (KiChjang) [inline components/style/servo/selector_parser.rs]: Because it shouldn't panic?

File: tests/wpt/metadata-css/selectors-3_dev/html/css3-selectors-lang-009.htm.ini
Comment (KiChjang) [inline tests/wpt/metadata-css/selectors-3_dev/html/css3-selectors-lang-009.htm.ini]: That's essentially https://tools.ietf.org/html/rfc4647#section-3.3.2, I believe. Will file after this is merged.

File: components/script/dom/element.rs
Comment (KiChjang) [inline components/script/dom/element.rs]: I'm not just not sure if that's possible or not.

Comment (KiChjang) [inline components/script/dom/element.rs]: I'm not sure if it's possible to have a non-elemnt node be a parent of an element. However your solution doesn't quite work, because we need some way of accessing the parent element of `current_element`, and there's no convenience method for this yet, meaning we'll need to upcast it back to a Node struct anyway just to call `parent_node_ref`.

File: components/script/dom/element.rs
Comment (KiChjang) [inline components/script/dom/element.rs]: Tried, didn't work. Type mismatch, and Attr cannot be dereferenced.

**Comment by KiChjang** [comment]
Comments addressed.

File: components/style/servo/selector_parser.rs
Comment (emilio) [inline components/style/servo/selector_parser.rs]: But it _is_ unreachable, since we do handle it right above this match. If it reaches this, it's a bug, so I'd like it to panic.

If you feel pretty strongly about it, at least `debug_assert!(false)`, or whatever, but...

File: components/style/servo/selector_parser.rs
Comment (emilio) [inline components/style/servo/selector_parser.rs]: How does it relate? It's effectively an immutable string, with `Box<str>` you can deref to `&str` and still use `split`, etc.

File: components/script/dom/element.rs
Comment (emilio) [inline components/script/dom/element.rs]: Oh, right, it's `AttrValue` the one that auto-derefs to string, never mind.

File: tests/wpt/metadata-css/selectors-3_dev/html/css3-selectors-lang-009.htm.ini
Comment (emilio) [inline tests/wpt/metadata-css/selectors-3_dev/html/css3-selectors-lang-009.htm.ini]: Why not filing it off-hand, and mark these (and your fixme) with that issue?

File: components/style/servo/selector_parser.rs
Comment (KiChjang) [inline components/style/servo/selector_parser.rs]: Oh, you can't even create a `Box<str>`:
```
error[E0277]: the trait bound `str: std::marker::Sized` is not satisfied
   --> /home/keith/Workspace/servo/components/style/servo/selector_parser.rs:252:28
    |
252 |             "lang" => Lang(Box::new(*try!(parser.expect_ident_or_string()))),
    |                            ^^^^^^^^ the trait `std::marker::Sized` is not implemented for `str`
    |
    = note: `str` does not have a constant size known at compile-time
    = note: required by `<std::boxed::Box<T>>::new`
```

File: tests/wpt/metadata-css/selectors-3_dev/html/css3-selectors-lang-009.htm.ini
Comment (KiChjang) [inline tests/wpt/metadata-css/selectors-3_dev/html/css3-selectors-lang-009.htm.ini]: Filed #15746.

File: components/style/servo/selector_parser.rs
Comment (emilio) [inline components/style/servo/selector_parser.rs]: I'm pretty sure you can if you do: `parser.expect_ident_or_string().into_owned().into_boxed_str()`: https://play.rust-lang.org/?gist=184a2e91d53c9bebfd852ec73ccc34a6&version=stable&backtrace=0

Comment (KiChjang) [inline components/style/servo/selector_parser.rs]: Wow, this is magic!

File: components/script/dom/element.rs
Comment (emilio) [inline components/script/dom/element.rs]: Let's use here `or_else` and then just one `map` to avoid the `String::from(attr.Value())` duplication.

File: components/script/dom/element.rs
Comment (emilio) [inline components/script/dom/element.rs]: I suggested this should be using `Element` itself, and doesn't seem addressed. If there's any particular reason it can't be done, or you think it's not worth it, please comment about it, otherwise, please do :)

**Review by emilio — APPROVED** [review]
r=me with those changes, thanks! :)

File: components/script/dom/element.rs
Comment (KiChjang) [inline components/script/dom/element.rs]: Perhaps it got lost from the GH notifications, but here was the original reply:
> I'm not sure if it's possible to have a non-element node be a parent of an element. However your solution doesn't quite work, because we need some way of accessing the parent element of current_element, and there's no convenience method for this yet, meaning we'll need to upcast it back to a Node struct anyway just to call parent_node_ref.

Comment (emilio) [inline components/script/dom/element.rs]: Oh, ok... I'm still kinda uneasy with the `None => continue`, but I can probably live with it?

**Pull request merged**
